### ISSUE TITLE
allow ones to override Record.hashCode

### ISSUE THREAD

**Issue by kojisekig** [NLP4L/framework#22]
Currently, all Cells of a Record are used to calculate hashCode of the Record. In some cases, we want not to take into account some of Cells that are used in a Record when calculating the Record.hashCode.

For example, there is a Processor for TermsExtractor has two cells, the term cell and score cell. Even though a term is same but its score can be different between multiple runs. In this case, the Record that was deleted by an operator at the previous run cannot be deleted by the ReplayProcessor at the next run because of different hashCode.

**Comment by yamamoto-takashi** [comment]
I implemented this feature as following example;

```scala
class SimpleDictionaryAttributeFactory(settings: Map[String, String]) extends DictionaryAttributeFactory(settings) {
  override def getInstance: DictionaryAttribute = {
    
    /**
     * User defined cell hashcode function
     */
    def ignoreThisCell(d: Any): Int = {
      0
    }
    
    val list = Seq[CellAttribute](
      CellAttribute("cell01", CellType.StringType, true, true),
      new SimpleCellAttribute("cell02", CellType.IntType, false, true, ignoreThisCell),
      CellAttribute("cell03", CellType.DoubleType, false, true, ignoreThisCell),
      CellAttribute("cell02_check", CellType.StringType, false, false, ignoreThisCell),
      CellAttribute("cell04", CellType.FloatType, false, true, ignoreThisCell)
      )
    new DictionaryAttribute("simple", list)
  }
}
```

I added the new parameter in CellAttribute which is a hashcode function (Any) => Int.
If you'd like to ignore a cell value in ReplayProcessor, then you set a function that returns static value such like zero.

If this implementation accepted, I will push the new branch.

**Comment by kojisekig** [comment]
Basically, I like the idea as it is very simple, if I understand correctly what you mean. However, the Cell that its ignore flag is turned on can still contribute the hashCode of the Record that the Cell participates in? For example, when a Record.hashCode is being calculated and its current value is 0x0000 and the next Cell.hashCode that is marked as ignored returns 0x0000, and if XOR is used in Record.hashCode to mix in, the value turns to 0xFFFF.

Nonetheless, ignored Cells always return same constant and same hashCodes are produced in the Records. Thinking of it, the hashCode of the ignored Cell can be any constant. I think the proposed way works.

I may be wrong. Can you push the branch so that I can see and try? Thanks!

**Comment by yamamoto-takashi** [comment]
Anyway, I push the branch " add_user_own_hashcode".
Please check it and report your comment. Thanks.

**Comment by kojisekig** [comment]
I added a test case and merged. Thanks!

**Comment by kojisekig** [comment]
Hi Yamamoto-san,

After I applied this change, I cannot see "MOD" flag is turned on in Replay field...

**Comment by yamamoto-takashi** [comment]
I fix this bug in bugfix_MOD_label_of_replay branch.
Could you check this branch.
Thanks.

**Comment by kojisekig** [comment]
I still see the problem even I use the branch you mentioned. I can see ADD flag is turned on but I cannot see MOD flag in the Replay field.

BTW, I applied this before trying your branch:

https://github.com/NLP4L/framework/commit/e65a9b4d442a53f58317f1353c1db149bcff3c81

What I've tried on the branch you mentioned was that I applied above patch to bugfix_MOD_label_of_replay and Run SimpleProcessor which is defined in examples/simple.conf.

Can you test the branch with examples/simple.conf?

**Comment by yamamoto-takashi** [comment]
I confirm this problem, however I successfully see MOD and ADD label after replay examples/simple.conf job.

These are my processure.

(1) Create a job using examples/simple.conf
(2) Run onne. Getting the #1 result
(3) Run again. Getting the #2 result.
(4) Delete the record of cell01 is 19 on #2 
(5) Modify the record of cell01 is 18 to 18--
(6) Add a  record of cell01 is aaaaa
(7) Run again. Getting the #3 result.
![mod_del_add](https://cloud.githubusercontent.com/assets/466845/12423935/27d3530e-bf11-11e5-9f75-f141d619218c.png)

**Comment by kojisekig** [comment]
Thank you very much for showing me details all you've done to confirm the problem with a nice screen capture! I realized that the record I modified was the second record which has "var1" in cell01, and it was merged with the first record which has the same value in cell01 by MergeProcessor. That's why I couldn't see any records that have MOD flag in Run 2.

When I make a change on another unique record, I can see the record has MOD flag.

I'll merge the branch soon!
